MSI install

tmeckel opened the pull request:
This pull request contains the following components

* an MSI package definition based on the WiX Toolset to create an MSI package for OpenSSH ([README.md](https://github.com/tmeckel/openssh-portable/blob/msi-install/contrib/win32/install/install/README.md))
* Build scripts that can be used standalone or a in CI/CD pipeline to build OpenSSH and/or the MSI Package ([README.md](https://github.com/tmeckel/openssh-portable/blob/msi-install/contrib/win32/install/build/README.md))
* Dockerfiles for creating docker images to build OpenSSH and/or the MSI Package ([README.md](https://github.com/tmeckel/openssh-portable/blob/msi-install/contrib/win32/install/docker/README.md))

tmeckel committed:
Initial check in of MSI infrastructure

tmeckel committed:
Fixed an error in build-msi.ps1 and set default for BUILD_TAG in Docker to head

tmeckel committed:
SolutionDir property will now be set in paths.targets if empty and set to $(ProjectDir) in that case to ensure a non empty SolutionDir property is present if projects aren't build using Visual Studio.

tmeckel committed:
Making sure that Start-OpenSSHPackage from OpenSSHBuildHelper.psm1 will always exit on an error.

tmeckel committed:
Removed special handling of missing moduli file from  build-openssh.ps1

tmeckel committed:
Corrected advertising flags in product.wxs

tmeckel committed:
Removed left-over (debug) message of var.SOURCEDIR from product.wxs

tmeckel committed:
Added documentation

SteveL-MSFT commented:
@heaths, can you help review the WIX files?  Thanks!

heaths commented on contrib/win32/install/build/README.md:
```diff
@@ -0,0 +1,43 @@
+- name: Install OpenSSH
+  win_package:
+    path: openssh.msi
+    product_id: '{0240359E-6A4C-4884-9E94-B397A02D893C}'
```
You should never hardcode ProductCode. These should be auto-generated by WiX (if using WiX) or your own GUID. Major upgrade servicing is the ideal servicing path for small-ish packages (MSPs are fraught with problems and should be avoided). See https://devblogs.microsoft.com/setup/serviceability/ for a primer, and links to more information. This is based on over 20 years with MSI, and 15 on the .NET and VS teams working on their installs.

tmeckel replied:
@heaths That's the way Ansible works and I guess other Configuration Management frameworks as well (except SCCM because it's a Microsoft Product 😏😀 ). You can discuss that the people who implemented it this way did it wrong, but from my point of view it's only consequent to do it so because technically the ProductCode uniquely identifies a MSI installation package (and the product version implicitly) and as far that I understood it over the years it compares to the tuple `(packagename,version)` in APT or RPM. When I misunderstood this completely than it's likely that the Ansible guys did it as well and the concept of the Product Code in MSI is hard to explain and to grasp.

So, in addition I've to admit that I always wondered why WiX creates a new ProductCode (and why this should be a good thing) at every build even if the version (contents) of the product didn't change and you only needed to build a new MSI because you had to correct a spelling error in a MSI dialog for example.  In that case the product code shouldn't change because the product, you deliver using the MSI, didn't change. From my experience in the field I know that you always make great friends telling everybody your MSI has now a new product code because you changed the GUI of the MSI 😁 😁 Don't get me wrong, I definitely have less experience with MSI  than you as I can see, but the issue with those changing product codes and MSI was, and constantly is, always a pain.

That's the reason why I tried to elaborate in the README.md why I approached the subject a little different and added some code in the `build-msi.ps1` script to generate the product code based on a seed value and the version string. This procedure is standardized in chapter [4.3: Algorithm for Creating a Name-Based UUID of RFC-4122](https://tools.ietf.org/html/rfc4122#section-4.3). That way you’ll get the same product code for the same product version and the algorithm guarantees that you can calculate the product code offline by only knowing the seed value and the version. This makes the product code predicable for every party in the game who’s responsible in delivering software. I know that upgrading MSI installations is totally different beast but having the ability to calculate the product code in a foreseeable way has many advantages from my point of view (see the Ansible example).

I hope I didn't step on your toes with my statements, as we Germans are known to communicate rather directly.

heaths replied:
Because keeping the same ProductCode when bumping the ProductVersion constitutes a minor upgrade which you can't simply install over top of another, e.g. double click. This is all documented in the MSI docs.

If you have the product installed, you need to *reinstall* a minor upgrade, which can only be done from the command line (e.g. `msiexec /i product.msi REINSTALL=ALL REINSTALLMODE=vomus`). Obviously, that's less than ideal for customers who just want to install it. Using a chainer like WiX's burn can help simplify that story, though. That also increases size and build complexity, though, and burn was primarily meant for installing multiple projects as a single, cohesive product.

tmeckel replied:
@heaths as I tried to explain above, when you create new new version (pass a new version to the RFC-4122 algorithm) the ProductCode will do change for sure. The only difference to your favorite solution allowing WiX to a create a new product code at every build, are two facts:

1. The ProductCode is predictable. What is great from my experience. IT should be predictable, right? Randomness has definitely it's justification, but not here for my taste.
2. The ProductCode will not change when you don't change the product version (e.g. fixing a type in the UI). WiX will ALWAYS create a new PrdoctCode on every run.

heaths commented on contrib/win32/install/build/build-msi.ps1:
```diff
@@ -0,0 +1,224 @@
+	/// <summary>
+	/// Helper methods for working with <see cref="Guid"/>.
+	/// </summary>
+	public static class GuidUtility
```
WiX properly generates stable GUIDs. You should use `Product/@Id="*"` for the ProductCode, and leave off the `Component/@Guid` attribute whenever possible, i.e. when you have a key path file or registry value. This was the major change I did for PowerShell itself to fix it's shared component and upgrade issues it had between stable and preview releases. Please don't repeat the same mistakes. WiX does all the hard work itself better.

tmeckel replied:
See comment above

heaths commented on contrib/win32/install/build/build-msi.ps1:
```diff
@@ -0,0 +1,224 @@
+	if (-not (Test-Path -Path $msBuildPath)) {
+		$msBuildPath = Join-Path -Path $msBuildDir -ChildPath 'MSBuild\15.0\Bin\MSBuild.exe'
+		if (-not (Test-Path -Path $msBuildPath)) {
+			if ($env:CI) {
```
This pattern for CI logging is repeated. Recommend breaking it out into a separate function. Should make the script easier to review and maintain.

tmeckel replied:
I had a look at the packaging stuff in the PowerShell Github repo. Perhaps we should replace the code here with the stuff from Powershell entirely.

tmeckel replied:
As I said I had a look at the PowerShell Modules used in the the PowerShell repo, which support the CI build. And there you'll find numerous places with the same pattern:

Like this

```Powershell
if ($env:TF_BUILD)
{
    Write-Host "##vso[artifact.upload containerfolder=wix;artifactname=wix]$newXmlFileName"
}

throw "Current files to not match  {$FilesWxsPath}"
```

So I'd consider my code not as bad as it might look.

heaths commented on contrib/win32/install/install/README.md:
```diff
@@ -0,0 +1,54 @@
+
+* `MANUFACTURER`: (optional) the manufacturer of the MSI; default: Microsoft Corporation
+* `CONFIGURATION`: (required) the configuration used; Valid values: `Release`, `Debug`
+* `VERSION`: (optional) the version tag of the MSI; default: `0.0.0.0`
```
This will not work for servicing - for people to install newer builds over older. The `Product/@Id="*"` is auto-GUID'd, and the `Product/@Version` should be fed in from, say, the current build number. This allows people to just download and "run" a newer MSI and have it upgrade. It doesn't get any easier than this. As you're doing it now requires re-installing the MSI.

tmeckel replied:
Then we should remove defaults entirely and ensure that the Wix pre-processor issues an error if the parameter is not set.

tmeckel replied:
Then we should remove the default values in `product.wxs` completely and ensure that the pre-processor creates an error if the parameter is not set. About the the product code I refer to my commet above.

tmeckel replied:
`Version` is now a required input to `product.wxs`

heaths commented on contrib/win32/install/install/README.md:
```diff
@@ -0,0 +1,54 @@
+
+### Upgrade Codes
+
+* Debug Version: `c7a00fb1-c477-40d3-a951-ec4c749da439`
```
These need to vary by bitness as well, if using different x86 and x64 MSIs (recommended). See what I did for PowerShell itself for ideas.

tmeckel replied:
Good point, but isn't Windows INTEL/AMD the only plattform which has this issue because there is this infamous WOW64 😁  and @SteveL-MSFT do we have to support 32bit Windows anyway?

heaths replied:
Windows Installer actually differentiates between x86 and x64, disabling WOW64 virtualization in almost all cases. You can't assume it with installer packages.

tmeckel replied:
Removed the debug version stuff

heaths commented on contrib/win32/install/install/components/c_client.wxi:
```diff
@@ -0,0 +1,11 @@
+<?xml version="1.0" encoding="utf-8"?>
+<Include>
+  <Component Id="C_CLIENT" Guid="579a4404-dfba-49c2-a9b6-6f6319cf7449">
```
Do not hardcode GUIDs when not necessary. This can break a number of scenarios. WiX generates stable GUIDs (i.e. across builds) given the known folder or registry paths.

Ideally, each file should have it's own component and when a component has one file, it will be the key path and used to genarate a stable GUID.

The only time one file per component is not advised is when they are unversioned, or there are a LOT. This MSI does not have a lot of files. This means files are upgraded correctly and, depending on `RemoveExistingProducts` is scheduled, efficiently.

heaths replied:
Same comments apply to all the WiX authoring below.

heaths commented on contrib/win32/install/install/components/c_client.wxi:
```diff
@@ -0,0 +1,11 @@
+    <File Id="sftp.exe" Source="$(var.SOURCEDIR)/sftp.exe" Name="sftp.exe" />
+    <File Id="sshadd.exe" Source="$(var.SOURCEDIR)/ssh-add.exe" Name="ssh-add.exe" />
+    <File Id="sshkeyscan.exe" Source="$(var.SOURCEDIR)/ssh-keyscan.exe" Name="ssh-keyscan.exe" />
+    <Environment Id="PATH" Name="PATH" Value="[INSTALLFOLDER]" Permanent="no" Part="first" Action="set" System="yes" />
```
This would require a `Component/@Guid` if in its own component, so put it in one of the versioned EXE components above, like with _ssh.exe_. It then becomes dependent on ssh.exe being installed or upgraded.

heaths commented on contrib/win32/install/install/components/c_client.wxi:
```diff
@@ -0,0 +1,11 @@
+<?xml version="1.0" encoding="utf-8"?>
+<Include>
```
Why use includes? You can use separate .wxs files and just link them all together. WiX does that more efficiently.

heaths commented on contrib/win32/install/install/openssh-install.wixproj:
```diff
@@ -0,0 +1,68 @@
+    <ProjectGuid>9607b2ee-4071-4e93-b0a5-a316c3b1ad44</ProjectGuid>
+    <SchemaVersion>2.0</SchemaVersion>
+    <OutputType>Package</OutputType>
+    <Version Condition=" '$(Version)' == '' ">0.0.0.0</Version>
```
This should be populated by the build system to support major upgrades.

tmeckel replied:
So should we emit an error if the Version is not specified? What about the other default values?

heaths commented on contrib/win32/install/install/product.wxs:
```diff
@@ -0,0 +1,170 @@
+  <?error Unsupported value of sys.BUILDARCH=$(sys.BUILDARCH)?>
+<?endif?>
+
+<?define UPGRADECODERELEASE = "c7a00fb1-c477-40d3-a951-ec4c749da439" ?>
```
These need to vary by bitness as well, which you'll need to pass in.

tmeckel replied:
Exactly... but this is only relevant if we really have to support 32bit with OpenSSH. Right?

heaths replied:
Yes, but are you suggesting that 32-bit Windows doesn't have OpenSSH? I honestly don't know - I haven't installed a 32-bit OS in over a decade.

tmeckel replied:
See that's my point as well. Never used 32-bit OS for years now. Perhaps we could say goodbye to this architecture here. @SteveL-MSFT what's you opinion on this? Do we have to support 32bit or not?

tmeckel replied:
Introduced upgrade codes for x86 and x64

heaths commented on contrib/win32/install/install/product.wxs:
```diff
@@ -0,0 +1,170 @@
+<?define UPGRADECODEDEBUG = "e2dd3c95-7a8b-444e-b75b-3e1cf697aadc" ?>
+
+<?if $(var.PRODUCTID) = "" ?>
+  <?define PRODUCTID = "*" ?>
```
I see you are changing it, so I don't understand the hardcoded product ID in the README above. Whatever the case, this won't work as expected unless the version is incremented with the build.

tmeckel replied:
Because I simply want to have the ability to go with "best practices" even if I don't like them 😁 in the "normal" build process the `build-msi.ps1` will feed the generated Product Code into Wix

heaths commented on contrib/win32/install/install/product.wxs:
```diff
@@ -0,0 +1,170 @@
+    <Property Id="LOGVERBOSE" Value="LOGVERBOSE"/>
+
+    <!-- https://stackoverflow.com/a/19664484 -->
+    <Upgrade Id="$(var.UPGRADECODEBLOCKED)">
```
This wouldn't be necessary if properly authored as a major upgrade.

tmeckel replied:
I added this here because I wanted to prevent people from installing a debug version over a release version with the same product version string. But if this not required at all, or can be achieved by using different product code or by other means this can be removed for sure.

heaths replied:
It just makes the UX harder. If they want to flip between retail and debug, that should be up to them.

tmeckel replied:
Well okay, I've to admit maybe I really overshot the mark. So I will remove that part here 👍🏼

tmeckel replied:
Removed!

heaths commented on contrib/win32/install/install/product.wxs:
```diff
@@ -0,0 +1,170 @@
+			<ComponentGroupRef Id="CG_Main" />
+      <ComponentGroupRef Id="CG_Main_Symbols"/>
+      <Feature Id="F_Client" Title="Client programs" Level="1" Absent="allow" AllowAdvertise="no">
+        <Condition Level="10000">DISABLE_CLIENT</Condition>
```
Why using conditions in features instead of just features. If you just want customers to be able to opt into what features they want installed, it's more intuitive - especiually in UI - to just use features.

On the command line, for example, features can be added (if not enable by default) or removed (if enabled by default) like so:

```powershell
msiexec /i openssh.msi 'ADDLOCAL=A;B' REMOVE=C
```

In the UI, you'll actually see a feature tree. The conditions here are extraneous. You could just mark the features as off by default by authoring their level less than the `INSTALLLEVEL` property which is, by default, 1. So if you just mark these as `@Level=2`, they will be off by default.

tmeckel replied:
To my shame, I must confess that I did not know this feature of MSIEXEC until now. So we definitly can remove those conditions and simply go with the bare features. But I've to say that I never had issues using the procedure with explicit conditions even in the UI.

heaths commented on contrib/win32/install/install/product.wxs:
```diff
@@ -0,0 +1,170 @@
+      <?include components/c_main.wxi ?>
+		</ComponentGroup>
+    <ComponentGroup Id="CG_Main_Symbols" Directory="INSTALLFOLDER">
+      <?include components/c_main_symbols.wxi ?>
```
Rather than includes, just use a `ComponentRef` or `ComponentGroupRef` like you did above for things defined here, but then just link those .wxs files in. Just add them as `<Compile>` items and WiX does the rest. It links them all together like a C++ project would link .obj files created from .cpp files - in fax, WiX build was modeled after this: .wxs -> .wixobj -> .msi + .wixpdb.

tmeckel replied:
I checked the WXS source code in the PowerShell GitHub repo and I understood the way to go. Makes the installer even simpler 👍🏼

tmeckel replied:
Removed the includes

heaths reviewed (changes requested):
It's a great start, but WiX can do a lot of this for you, including support major ugprades properly, even across different CPU architectures. THat's why it's vital you let WiX auto-GUID the ProductCode, use a per-arch + per-configuration UpgradeCode, and do not hardcode component GUIDs, which will create problems in some cases.

heaths commented:
@tmeckel if you'd like, I could help make some of these changes to your PR if you'd allow it. There's a lot of great scaffolding here for build scripts. Or, if easier and fine with @SteveL-MSFT, this could be merged and I could work on it afterward before anything is released (i.e. before the scripts are invoked, so maybe they could be unhooked from the build till then).

tmeckel commented on contrib/win32/install/install/components/c_server.wxi:
```diff
@@ -0,0 +1,110 @@
+      Stop="uninstall"
+      Remove="uninstall"
+      Wait="no" />
+    <fw:FirewallException
```
The firewall will only be opened when you install the server component and to me it makes barely any sense to run an OpenSSH server behind a closed firewall (okay, except for local testing). But perhaps this can be moved into it's own separate feature.

tmeckel replied:
Would be nice... But I don't believe this gonna happen.

tmeckel replied:
No.. The SSH client programs feature is the default

tmeckel replied:
Do you know how the Windows Server SSH Server capability behaves at that point? We should follow it's behavior to be consistent.  If this will be officially released by Microsoft

tmeckel committed:
Changed installer after first review

tmeckel committed:
Updated MSI documentation in contrib\win32\install\install\README.md

tmeckel commented:
@heaths I tried to follow all your advises and 

* replaced the conditionals by features
* the `Version` is now a required input to `product.wxs` and I removed the default values from `product.wxs` and `openssh-install.wixproj`
* each file is now in a component
* introduced upgradecodes for x86 and x64 (including the possibility to build preview versions)
* replaced the include files with "normal" Wix `<Fragments>` which will be liked together by `light`
* completely removed the handling of debug versions

The documentation has been updated accordingly.

@TravisEz13 I introduced a new MSI feature which implements the firewall exception for SSH. The feature is disabled by default.

![image](https://user-images.githubusercontent.com/14177833/109425840-ac99c980-79ea-11eb-9f8d-be531c552982.png)

tmeckel committed:
Removed default value for Version property in contrib\win32\install\install\openssh-install.wixproj

tmeckel committed:
Parameters ProjectPath and SourceDirectory are now mandatory in contrib\win32\install\build\build-msi.ps1

bagajjal commented on contrib/win32/install/build/README.md:
```diff
@@ -0,0 +1,43 @@
+# Windows OpenSSH - Build Scripts
+
+## `build-openssh.ps1`
```
We already have [openssh-build.ps1](https://github.com/PowerShell/openssh-portable/blob/latestw_all/contrib/win32/openssh/OpenSSH-build.ps1) which does the build and package.

Please modify this.

bagajjal commented on contrib/win32/install/build/build-msi.ps1:
```diff
@@ -0,0 +1,224 @@
+	[Parameter(Mandatory)]
+	[ValidateScript( { Test-Path -LiteralPath $_ -PathType Container })]
+	[string]
+	$SourceDirectory,
```
We already have a build pipeline which does compilation, sign all artifacts and finally creates a zip payload.
If I want to add one more task to create the MSI package then I just need to call this script with $projectpath as openssh-install.wixproj, $sourcedirectory as <artifact_bin_folder>, $destination as <dest_folder_for_final_msi>, $version is derived from the executables and platform/configuration?

Do I need to change any GUIDs? I saw there is some discussion about hardcoding GUID vs generating GUID automatically.

bagajjal commented on contrib/win32/install/build/build-openssh.ps1:
```diff
@@ -0,0 +1,136 @@
+[CmdletBinding()]
```
We already have [openssh-build.ps1](https://github.com/PowerShell/openssh-portable/blob/latestw_all/contrib/win32/openssh/OpenSSH-build.ps1)  which does the build and package.

Please remove this.

bagajjal commented on contrib/win32/install/docker/Dockerfile-build:
```diff
@@ -0,0 +1,65 @@
+# escape=`
+
+# Copyright (C) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT license. See LICENSE.txt in the project root for license information.
```
We also have docker [based build](https://github.com/PowerShell/openssh-portable/blob/latestw_all/contrib/win32/openssh/vstsbuild.ps1).

This can be removed.

bagajjal commented on contrib/win32/install/install/README.md:
```diff
@@ -0,0 +1,63 @@
+
+#### x64 (64-bit)
+
+* Preview version: `5a388d64-a109-4d85-bddc-7f46e1e1e520`
```
What's the importance of these GUIDs? Given these are hardcoded, is there a need for changing them when a new OpenSSH version is published?

bagajjal commented on contrib/win32/install/install/components/c_client.wxs:
```diff
@@ -0,0 +1,25 @@
+      <Component Id="cmp_sshkeyscan.exe">
+        <File Id="sshkeyscan.exe" KeyPath="yes" Source="$(var.SOURCEDIR)/ssh-keyscan.exe" Name="ssh-keyscan.exe" />
+      </Component>
+      <Component Id="cmd_sshclient_env" KeyPath="yes" Guid="d71ba15b-3489-4b44-a89f-76b147c574a6">
```
Can we make this optional? We can have it enabled by default but should give an option for end-user to uncheck the option if user don't want to alter the PATH.

bagajjal commented on contrib/win32/install/install/components/c_server.wxs:
```diff
@@ -0,0 +1,138 @@
+          <util:EventManifest />
+        </File>
+        <util:XmlFile
+          Id="xmlSetResourceFileName"
```
Can you educate me on what we are doing here?

heaths replied:
This modifies the `resourceFileName` to point at the actual installed executable since it's hardcoded in the manifest. It's a great catch, but the `ResourceFile` attibute of `util:EventManifest` will do that automatically:

> The resource file (including path) of all the providers in the event manifest.  Often the resource file path cannot be determined until setup time. Put your  ResourceFile here and the resourceFileName attribute of the all the providers in  the manifest will be updated with the path before it is registered.

bagajjal commented on contrib/win32/install/install/components/c_server.wxs:
```diff
@@ -0,0 +1,138 @@
+            FirstFailureActionType="restart"
+            SecondFailureActionType="restart"
+            ThirdFailureActionType="restart" />
+          <PermissionEx Sddl="D:(A;;CCLCSWRPWPDTLOCRRC;;;SY)(A;;CCDCLCSWRPWPDTLOCRSDRCWDWO;;;BA)(A;;CCLCSWLOCRRC;;;IU)(A;;CCLCSWLOCRRC;;;SU)(A;;RP;;;AU)" />
```
What's the reason for providing "WriteExtendedAttributes" permissions for the "NT AUTHORITY\Authenticated Users"?
"NT AUTHORITY\SYSTEM" -  doesn't have all permissions. any reason why?
"BUILTIN\Administrators" - is it the full control?

![image](https://user-images.githubusercontent.com/23668037/114613315-04695880-9c58-11eb-91a0-031cb42b616d.png)

bagajjal replied:
What's the reason for setting the ACLs for ssh-agent service but not for SSHD service?

heaths replied:
As for the SDDL, this is copied directly from install-sshd.ps1.

heaths replied:
Also, install-sshd.ps1 doesn't set the ACL on sshd. Should it?

bagajjal replied:
This is to allow non-admin users to start the ssh-agent. ssh-agent is used for storing the keys. It's part of both client and server components.  Please note, non-admin users are not given permissions to stop the ssh-agent.

The same ACL is not present on sshd as we don't want the non-admin users to start the sshd.exe.

bagajjal commented on contrib/win32/install/install/components/c_server.wxs:
```diff
@@ -0,0 +1,138 @@
+      </Component>
+    </ComponentGroup>
+    <ComponentGroup Id="CG_SERVER_SSHDSERVICE" Directory="INSTALLFOLDER">
+      <Component Id="cmp_sshd.exe">
```
Can you add a check to make sure the sshd.exe binary has "read & execute" permissions for the "authenticated users".
Not sure if you are setting the OpenSSH folder permissions explicitly. If not openssh bin folder inherits the folder permissions from it's parent folder.. If the parent folder doesn't have "read & execute" permissions for "authenticated users" then ssh connection will fail.

tmeckel commented:
If I had realized how much work I've to put in this pull request, I would never have submitted it. I apologize for taking up your time.

floh96 commented:
@tmeckel sad that you closed it a msi installer for openssh would be a great addition!
@heaths @bagajjal would you consider continuing this pr?

heaths commented:
I will consider it. @SteveL-MSFT what scenarios would this target, exactly? Is it only for standalone installs, or would this facilitate MU updates some day, potentially? (I.e. should this include a Burn installer.)

@tmeckel I'm sorry you feel that way, but I'm happy to help more hands on if you want. MSI installations have a lot of rules not often understood, and they can have major ramifications that can potentially block servicing packages later. That is one major reason I architected them almost entirely out of the Visual Studio 2017 and newer installs.

tmeckel commented:
@heaths no it's definitely not about MSI or something technically. Please don't make me out to be a little kid. It's about how this PR has been handled. I should've been warned from the outcome of the first PR I filed here and how this ended.

https://github.com/PowerShell/openssh-portable/pull/399

@SteveL-MSFT Would be nice, if you guys would give me some credit, if you build your own stuff based on my contributions.

SteveL-MSFT commented:
@heaths I think we only want to support this as standalone installs of preview versions of OpenSSH.  Official installs will still be via feature-on-demand within Windows.

@tmeckel I appreciate you starting this effort!  Whether Heath builds upon this work or creates it from scratch, you do deserve credit for initiating this.

heaths commented on contrib/win32/install/install/components/c_server.wxs:
```diff
@@ -0,0 +1,138 @@
+        <Condition>
+          <![CDATA[NOT(DONT_START_SSHAGENT OR DONT_START_SERVICES)]]>
+        </Condition>
+        <ServiceControl
```
@bagajjal @SteveL-MSFT any reason to conditionally start the service?

bagajjal replied:
Yes. sshd_config (sshd serer configuration) can be configured to be more secured based on the users need,
1) changing default port from 22 to user specific port to avoid security attacks.
2) An admin user can configure which users are allowed to login (Allowusers, DenyUsers)
3) An admin user can configure the type of access like providing only sftp access and denying interactive shell, etc.

heaths replied:
They can edit the file and restart the service immediately after install. Is putting a condition on this something you think people will actually use? Does it need to be exposed in the UI that most admins will probably use?

/cc @SteveL-MSFT

bagajjal replied:
As part of MSI, I would give an option (disabled by default, to match existing install-sshd.ps1 behavior) to start the sshd, ssh-agent service. This gives control for admin users who want to first make changes to sshd_config before starting the sshd service.

heaths commented:
What's disabled by default, exactly? The option or the services being started? IMO, starting the service by default makes sense - most people expect an installer to install the product fully on their machine in a working state.

bagajjal commented:
Sure. We should provide a checkbox for users. If the checkbox is checked then only we should start the service. 
The checkbox can be checked by default.
